### ISSUE TITLE
Realm: play nicely with pre-existing signal handlers

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#1559]
It would be nicer of Realm if, when installing the `realm_backtrace` signal handler, it also saved any existing signal handler. At signal handler invocation the handler would unregister itself, restore the previous handler, and re-raise the same signal after printing out its message. This gives the pre-existing handler a chance to handle the signal.

This is e.g. what `cpython` does, when asked to install its own python-level-backtrace-printing signal handler (enabled with `PYTHONFAULTHANDLER=1`).

**Comment by eddy16112** [comment]
I did some experiment by caching pre-existing handlers before registering the realm signal handler. However, It turns out that GASNet has already override any user-defined handlers with its own signal handlers during GASNet initialization. We can cache the signal handlers before initializing the network, such that we should be able to catch the user-defined handlers. However, we will lose the gasnet error handler such as `GASNET_BACKTRACE=1`, maybe it is OK since we already have the `LEGION_BACKTRACE`.

**Comment by lightsighter** [comment]
Nobody should be using `LEGION_BACKTRACE`. That is a crutch that Realm provides. Everybody should be using `REALM_BACKTRACE` and `REALM_FREEZE_ON_ERROR`. I would focus on providing support for those.

**Comment by eddy16112** [comment]
> Nobody should be using `LEGION_BACKTRACE`. That is a crutch that Realm provides. Everybody should be using `REALM_BACKTRACE` and `REALM_FREEZE_ON_ERROR`. I would focus on providing support for those.

Yeah, I mean `REALM_BACKTRACE`. If it is OK to not have the support of `GASNET_BACKTRACE=1`, we can cache whatever application defined signal handlers before initializing the network, otherwise, we will ask GASNET folks to add the support of restoring any pre-existing signal handlers.

**Comment by lightsighter** [comment]
If I set `REALM_BACKTRACE=1` today, doesn't that already supplant the GASNet signal handlers?

**Comment by eddy16112** [comment]
> If I set `REALM_BACKTRACE=1` today, doesn't that already supplant the GASNet signal handlers?

Yes, I just want to make sure that we do not need any GASNet signal handlers once `REALM_BACKTRACE=1` or `REALM_FREEZE_ON_ERROR` is set.

**Comment by lightsighter** [comment]
I think that is already the case today, but please check on that with @streichler.

**Comment by eddy16112** [comment]
I just realized that even without `REALM_BACKTRACE=1`, we can not restore the pre-existing signal handlers, because GASNet will always override the signal handles. To fix it, we will need to do something like this:
```
  save pre-existing signal handlers();
  init network();
  if (REALM_BACKTRACE is not set) {
    restore pre-existing signal handlers();
  }
```
Personally, I do not like this fix. A cleaner and easier fix is to tell GASNet not register its own handlers unless a GASNet env (e.g. GASNET_BACKTRACE) is set. I think that's a question for @streichler

**Comment by elliottslaughter** [comment]
@PHHargrove might have opinions on how to get GASNet to play nice here.

**Comment by manopapad** [comment]
Arguably even when GASNet is asked to install its signal handler, it should similarly save any pre-existing signal handlers, and restore them if its own signal handler gets invoked (and re-raise).

**Comment by PHHargrove** [comment]
I apologize for this terse reply.  I have a proposal deadline on Tue

- Have a look at `GASNET_NO_CATCH_SIGNAL` in GASNet-EX's README for a means to prevent us from installing specific handlers.
- Regarding chaining the previous handler: please consider entering an enhancement request bug at https://gasnet-bugs.lbl.gov .

**Comment by eddy16112** [comment]
I have created a PR for this feature https://gitlab.com/StanfordLegion/legion/-/merge_requests/960
